File: backend/app/schemas/analytics.py
from datetime import datetime

from pydantic import BaseModel


class SankeyNode(BaseModel):
    id: str
    name: str
    color: str | None = None
    value: int | None = None  # Explicit value for nodes without incoming links


class SankeyLink(BaseModel):
    source: str
    target: str
    value: int


class SankeyData(BaseModel):
    nodes: list[SankeyNode]
    links: list[SankeyLink]


class HeatmapDay(BaseModel):
    date: str
    count: int


class HeatmapData(BaseModel):
    days: list[HeatmapDay]
    max_count: int


class AnalyticsKPIsResponse(BaseModel):
    total_applications: int
    interviews: int
    offers: int
    application_to_interview_rate: float
    response_rate: float
    active_opportunities: int


# Interview Rounds Analytics Schemas


class FunnelData(BaseModel):
    round: str
    count: int
    passed: int
    conversion_rate: float


class OutcomeData(BaseModel):
    round: str
    passed: int
    failed: int
    pending: int
    withdrew: int


class TimelineData(BaseModel):
    round: str
    avg_days: float


class RoundProgress(BaseModel):
    round_type: str
    outcome: str | None
    completed_at: datetime | None
    days_in_round: int | None


class CandidateProgress(BaseModel):
    application_id: str
    candidate_name: str
    role: str
    rounds_completed: list[RoundProgress]
    current_status: str


class InterviewRoundsResponse(BaseModel):
    funnel_data: list[FunnelData]
    outcome_data: list[OutcomeData]
    timeline_data: list[TimelineData]
    candidate_progress: list[CandidateProgress]

File: frontend/src/lib/analytics.ts
import api from './api';

export interface SankeyNode {
  id: string;
  name: string;
  color?: string;
}

export interface SankeyLink {
  source: string;
  target: string;
  value: number;
}

export interface SankeyData {
  nodes: SankeyNode[];
  links: SankeyLink[];
}

export interface HeatmapDay {
  date: string;
  count: number;
}

export interface HeatmapData {
  days: HeatmapDay[];
  max_count: number;
}

export async function getSankeyData(): Promise<SankeyData> {
  const response = await api.get('/api/analytics/sankey');
  return response.data;
}

export async function getHeatmapData(year?: number | 'rolling'): Promise<HeatmapData> {
  const params: Record<string, string | number | boolean> = {};
  if (year === 'rolling') {
    params.rolling = true;
  } else if (year) {
    params.year = year;
  }
  const response = await api.get('/api/analytics/heatmap', { params });
  return response.data;
}

// Interview Rounds Analytics Types

export interface FunnelData {
  round: string;
  count: number;
  passed: number;
  conversion_rate: number;
}

export interface OutcomeData {
  round: string;
  passed: number;
  failed: number;
  pending: number;
  withdrew: number;
}

export interface TimelineData {
  round: string;
  avg_days: number;
}

interface RoundProgress {
  round_type: string;
  outcome: string | null;
  completed_at: string | null;
  days_in_round: number | null;
}

interface CandidateProgress {
  application_id: string;
  candidate_name: string;
  role: string;
  rounds_completed: RoundProgress[];
  current_status: string;
}

interface InterviewRoundsResponse {
  funnel_data: FunnelData[];
  outcome_data: OutcomeData[];
  timeline_data: TimelineData[];
  candidate_progress: CandidateProgress[];
}

export async function getInterviewRoundsData(
  period: string = 'all',
  roundType?: string
): Promise<InterviewRoundsResponse> {
  const params: Record<string, string> = { period };
  if (roundType) params.round_type = roundType;
  const response = await api.get('/api/analytics/interview-rounds', { params });
  return response.data;
}
